src/GraphQL/Federation/Resolvers/EntityResolver.cs

```diff
@@ -0,0 +1,213 @@
+            {
+                array[i] = Deserialize(fieldName, listGraphType.ResolvedType!, elementType, array[i]);
+            }
+            // convert the array to the intended list type
+            return listConverter.Convert(array);
+        }
+
+        // handle scalar graph types
+        if (graphType is ScalarGraphType scalarGraphType)
+            return scalarGraphType.ParseValue(value);
+
+        // handle object graph types
+        if (graphType is IObjectGraphType objectGraphType)
+        {
+            if (value is not Dictionary<string, object?> dic)
+                throw new InvalidOperationException($"The field '{fieldName}' is an object graph type but the value is not a dictionary");
+
+            return ToObject(valueType, objectGraphType, dic);
+        }
+
+        // union and interface types are not supported
+        throw new InvalidOperationException($"The field '{fieldName}' is not a scalar or object graph type.");
+    }
+
+    private record RepresentationDataLoader(IResolveFieldContext Context, Representation Representation) : IDataLoaderResult
+    {
+        public Task<object?> GetResultAsync(CancellationToken cancellationToken = default) => Representation.Resolver.ResolveAsync(Context, Representation.Value).AsTask();
+    }
+
+    /// <inheritdoc/>
+    public ValueTask<object?> ResolveAsync(IResolveFieldContext context)
+    {
+        // require the representations argument to be converted to the proper type before hitting this code
+        // e.g.: representationArgument.Parser += (value) => EntityResolver.Instance.ConvertRepresentations(schema, (System.Collections.IList)value);
+        var representations = (IEnumerable<Representation>)context.Arguments![REPRESENTATIONS_ARGUMENT].Value!;
+
+        // now that the representations have been validated, we can use them to resolve the entities using
+        //   the resolvers provided by the representations
+
+        // note: context.Copy is implicit due to the returned object being a list; otherwise it would be necessary,
+        //   as the context is referenced within a delegate passed to the SimpleDataLoader (see below)
+        //context = context.Copy();
+
+        var results = new List<RepresentationDataLoader>();
+        foreach (var representation in representations)
+        {
+            // using a data loader here causes the resolvers to run in serial or parallel based on the selected execution strategy.
+            // unfortunately this requires extra allocations whereas if the strategy was known this code could be optimized by
+            // either awaiting each resolver or collecting them and performing WaitAll. Note that this code counts on the fact
+            // that the context instance will not be reused due to a list being returned from this method.
+            results.Add(new RepresentationDataLoader(context, representation));
+        }
+
+        return new(results);
+    }
+}
```

Comment:
I wish I were able to write such a clean and readable code 😅
Well done!

Reply:
Thanks!  I was going to try to explain the code to make for an easier review, and I thought, well that just means the code isn't readable enough.  lol